### ISSUE TITLE
afterMapZoom called before map zoom

### ISSUE THREAD

**Issue by russellquinn** [mapbox/DEPRECATED-mapbox-ios-sdk#90]
It seems like when you do a programatic animated zoom, beforeMapZoom and afterMapZoom are BOTH called before the map has zoomed.

e.g.

[mapView zoomWithLatitudeLongitudeBoundsSouthWest:southWest northEast:northEast animated:YES];

**Comment by incanus** [comment]
After waaaaaaaay too much time playing around with these delegate methods, I'm considering removing them altogether and just making available `-mapViewRegionDidChange:`, much like MapKit. I think this would still allow for the same sorts of observations, albeit the caller would have to compare `zoom` and `centerProjectedPoint` before & after each call. The caller could also implement `performSelector:withObject:afterDelay:` in response as well as successive `cancelPerformSelector:target:argument:` in order to avoid duplicate calls. 

The reason for this, aside from not quite getting all methods to work, is a ton of code complexity around handling due to many possible scenarios: 
- user pans the map (move)
- user zooms the map with pinch (zoom)
- user zooms the map with two-finger tap (zoom)
- user zooms the map with one-finger double-tap (zoom)
- location services pans the map (move)
- location services zooms the map (zoom)
- location services pans and zooms the map (move & zoom)
- caller programmatically zooms the map with animation & center changes (zoom & move)
- caller programmatically zooms the map with animation & center does not change (zoom)
- caller programmatically zooms the map without animation & center changes (zoom & move)
- caller programmatically zooms the map without animation & center does not change (zoom)
- caller programmatically pans the map with animation (move)
- caller programmatically pans the map without animation (move)

All of these additionally have to check each other to determine whether the user initiated either/both actions by examining `UIScrollView` tracking and/or `UIGestureRecognizer` states (two different gestures that I can gather, not counting the taps), as well as whether deceleration is enabled for the map view due to the delays involved. 

In short, I can see why Apple doesn't provide these :-)

Any thoughts?

**Comment by russellquinn** [comment]
This might cause me big problems. Let me check.

**Comment by incanus** [comment]
It occurs to me that I can also try this approach internally, so I will give that a shot today as well.

**Comment by russellquinn** [comment]
Okay, yeah the expanded delegate callbacks are actually one of the great things about MapBox ;)

**Comment by incanus** [comment]
Ok, this has been completely rebuilt. Please let me know what you think. As a result, it was difficult (or non-intuitive at least) as to how to pass the "user initiated action" flag with the `mapViewRegionDidChange:` callback, so I removed that argument and reverted to the old version. It seems to me that it still serves a useful purpose: it mirrors the MapKit API, plus if the dev wants more specific info, they can use the before/after move/zoom callbacks, which do have that info and should bookend the actual region change calls.

**Comment by russellquinn** [comment]
Works good! Thanks.
